### PR TITLE
[Hybrid Parallel] Add Topology for hybrid communicate

### PR DESCRIPTION
**Pull request by ForFishes**
### PR types

New features
### PR changes

Others
### Describe

[Hybrid Parallel] Add Topology for hybrid communicate

### CODE REVIEW

File: python/paddle/distributed/fleet/base/topology.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+__all__ = ['CommunicateTopology', 'HybridCommunicateGroup']
+
+
+class CommunicateTopology(object):
+    def __init__(self, hybrid_names, dims):
+        self._parallel_names = hybrid_names
+        self._dims = dims
+        self.coordinate = collections.namedtuple('Coordinate',
+                                                 self._parallel_names)
+        self._word_size = reduce(lambda x, y: x * y, self._dims)
```
Comment (hutuxian) [inline python/paddle/distributed/fleet/base/topology.py]: `word` or `world`?

Comment (ForFishes) [reply python/paddle/distributed/fleet/base/topology.py]: fix it

File: python/paddle/distributed/fleet/base/topology.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+import paddle
+import collections
+import numpy as np
+from itertools import product
+from functools import reduce
+__all__ = ['CommunicateTopology', 'HybridCommunicateGroup']
+
+
+class CommunicateTopology(object):
```
Comment (sandyhouse) [inline python/paddle/distributed/fleet/base/topology.py:23]: Add a comment to describe what it is?

File: python/paddle/distributed/fleet/base/topology.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    def __init__(self, hybrid_names, dims):
```
Comment (sandyhouse) [inline python/paddle/distributed/fleet/base/topology.py]: hybrid_names or group_names?

File: python/paddle/distributed/fleet/base/topology.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        self._parallel_names = hybrid_names
```
Comment (sandyhouse) [inline python/paddle/distributed/fleet/base/topology.py]: or group_names

File: python/paddle/distributed/fleet/base/topology.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        self._world_size = reduce(lambda x, y: x * y, self._dims)
+
+        ranges = [range(d) for d in self._dims]
+        all_coordinate = [self.coordinate(*x) for x in product(*ranges)]
+
+        self._coord2rank = dict(zip(all_coordinate, range(len(all_coordinate))))
+        self._rank2coord = dict(
+            zip(self._coord2rank.values(), self._coord2rank.keys()))
+
+    def get_parallel_names(self):
```
Comment (sandyhouse) [inline python/paddle/distributed/fleet/base/topology.py]: as above

File: python/paddle/distributed/fleet/base/topology.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    def get_axis_list(self, axis_name, index):
+        axis = self._parallel_names.index(axis_name)
+        ranks = [
+            self._coord2rank[coord] for coord in self._coord2rank.keys()
+            if coord[axis] == index
+        ]
+        ranks.sort()
+        return ranks
+
+    def get_dim_num(self, axis_name):
```
Comment (sandyhouse) [inline python/paddle/distributed/fleet/base/topology.py]: get_dim_size?

**Review by sandyhouse — DISMISSED** [review]
LGTM for others

File: python/paddle/distributed/fleet/base/topology.py (hunk 2 above)
Comment (ForFishes) [reply python/paddle/distributed/fleet/base/topology.py:23]: done

**Review by XieYunshen — APPROVED** [review]

**Review by sandyhouse — APPROVED** [review]
LGTM

**Review by fuyinno4 — APPROVED** [review]

**Pull request merged**